File: apps/backend/workflow_engine/workflow_engine/models/node_template.py
"""
Node Template Database Model.
"""

from sqlalchemy import Column, String, Boolean, JSON, text
from sqlalchemy.dialects.postgresql import UUID, ARRAY
import uuid

from .database import Base

class NodeTemplate(Base):
    __tablename__ = 'node_templates'

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    template_id = Column(String, unique=True, nullable=False)
    name = Column(String, nullable=False)
    description = Column(String)
    category = Column(String)
    node_type = Column(String, nullable=False)
    node_subtype = Column(String, nullable=False)
    version = Column(String, default='1.0.0')
    is_system_template = Column(Boolean, default=False)
    default_parameters = Column(JSON)
    required_parameters = Column(ARRAY(String))
    parameter_schema = Column(JSON) 

File: apps/frontend/agent_team_web/src/types/node-template.ts
// Node template types based on node-template.json

export type NodeCategory =
  | 'Trigger'
  | 'AI Agents'
  | 'Actions'
  | 'External Integrations'
  | 'Flow Control'
  | 'Human Interaction'
  | 'Memory'
  | 'Tools';

export type NodeTypeEnum =
  | 'TRIGGER'
  | 'AI_AGENT'
  | 'ACTION'
  | 'EXTERNAL_ACTION'
  | 'FLOW'
  | 'HUMAN_IN_THE_LOOP'
  | 'MEMORY'
  | 'TOOL';

export interface ParameterSchema {
  type: 'object';
  properties?: Record<string, SchemaProperty>;
  required?: string[];
}

export interface SchemaProperty {
  type: 'string' | 'boolean' | 'integer' | 'number' | 'array' | 'object';
  description?: string;
  enum?: string[];  // Dropdown options (from backend "options" field, converted to JSON Schema "enum")
  default?: unknown;

  // UI/Behavior properties
  sensitive?: boolean;      // Password field with masking
  multiline?: boolean;      // Textarea instead of input
  readonly?: boolean;       // Non-editable display field

  // Validation properties
  min?: number;             // Min value for numbers, min length for strings
  max?: number;             // Max value for numbers, max length for strings
  validation_pattern?: string; // Regex pattern for validation

  // Dynamic dropdown properties
  api_endpoint?: string;    // API URL to fetch dropdown options
  search_endpoint?: string; // API URL to search and fetch options (for searchable fields)
  multiple?: boolean;       // Multi-select for arrays

  // UI enhancement properties
  placeholder?: string;     // Input placeholder text
  help_text?: string;       // Help text/tooltip content

  // Nested schemas
  items?: SchemaProperty;
  properties?: Record<string, SchemaProperty>;
}

export interface NodeTemplate {
  id: string;
  name: string;
  description: string;
  node_type: NodeTypeEnum;
  node_subtype: string;
  version: string;
  is_system_template: boolean;
  default_parameters: Record<string, unknown>;
  required_parameters: string[] | null;
  parameter_schema: ParameterSchema;
  input_params: ParameterSchema | Record<string, unknown>;  // Input parameters schema
  output_params: ParameterSchema | Record<string, unknown>; // Output parameters schema
}

export interface NodeTemplatesData {
  node_templates: NodeTemplate[];
}
